Add InitialDelay to CashTrickler and add CashTricklerBar

MustaphaTR opened the pull request:
Fixes  #14022 and also adds a CashTricklerBar trait.

MustaphaTR committed:
Add InitialDelay to CashTrickler

pchote commented on OpenRA.Mods.Common/Traits/CashTrickler.cs:
```diff
@@ -21,6 +21,9 @@ public class CashTricklerInfo : ConditionalTraitInfo
 		public readonly int Interval = 50;
 
+		[Desc("Number of ticks to wait before giving first money.")]
+		public readonly int InitialDelay = -1;
```
0 would be a better default here.

pchote commented on OpenRA.Mods.Common/Traits/CashTrickler.cs:
```diff
@@ -37,12 +40,13 @@ public class CashTrickler : ConditionalTrait<CashTricklerInfo>, ITick, ISync, IN
 		readonly CashTricklerInfo info;
 		PlayerResources resources;
-		[Sync] int ticks;
+		[Sync] public int Ticks;
```
Please make this a read-only property so that other code can't modify it.

pchote commented on OpenRA.Mods.Common/Traits/Render/CashTricklerBar.cs:
```diff
@@ -0,0 +1,58 @@
+		{
+			this.self = self;
+			this.info = info;
+			cashTrickler = self.TraitsImplementing<CashTrickler>().ToArray();
```
Use `readonly IEnumerable<CashTricker> enabledTrickers;` and `enabledTrickers = self.TraitsImplementing<CashTrickler>().ToArray().Where(ct => !ct.IsTraitDisabled)` to avoid creating unnecessary objects.

pchote commented on OpenRA.Mods.Common/Traits/Render/CashTricklerBar.cs:
```diff
@@ -0,0 +1,58 @@
+
+		float ISelectionBar.GetValue()
+		{
+			var firstEnabledCashTrickler = cashTrickler.FirstOrDefault(ct => !ct.IsTraitDisabled);
```
You can then also use `var complete = enabledTricklers.MaxBy(ct => (float)ct.Ticks / ct.Info.Interval);` and `return 1 - complete;` to show the bar for the next trickle instead of the first.

MustaphaTR commented:
Updated.

MustaphaTR committed:
Make CashTrickler PausableConditional

MustaphaTR committed:
Add CashTricklerBar

MustaphaTR commented:
Updated.

Previous code were crashing when all CashTricklers were disabled. I changed CashTrickler to Pausable conditional, disabling it now resets Ticks=, pausing doesn't. CashTricklerBar no longer checks for IsTraitDisabled, as it would return to 0 anyway as Ticks = Interval when CashTrickler is disabled.

pchote commented on OpenRA.Mods.Common/Traits/Render/CashTricklerBar.cs:
```diff
@@ -0,0 +1,59 @@
+
+		float ISelectionBar.GetValue()
+		{
+			if (cashTricklers == null)
```
This will never be null, so can be removed.

pchote commented on OpenRA.Mods.Common/Traits/Render/CashTricklerBar.cs:
```diff
@@ -0,0 +1,59 @@
+				return 0;
+
+			var complete = cashTricklers.Min(ct => (float)ct.Ticks / ct.Info.Interval);
+			return 1 - complete;
```
`ct.Ticks / ct.Info.Interval` is always between 0 and 1, so this can be simplified to `return cashTricklers.Max(ct => (float)ct.Ticks / ct.Info.Interval);`

pchote reviewed (dismissed):
Works as advertised, and code looks good aside from a couple of minor nits:

reaperrr commented:
[changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/5238da042df49dcd789c227d57fbcc008aaf6b66...9c9e8b5d42cdda7dbb87f4c7e2e7856969de2f81)
